core-backend/common/src/lib.rs

```diff
@@ -24,23 +24,59 @@ extern crate alloc;
 
 pub mod funcs;
 
-use alloc::{boxed::Box, collections::BTreeMap};
+use alloc::{boxed::Box, collections::BTreeMap, string::String, vec::Vec};
 use gear_core::{
     env::Ext,
+    gas::GasAmount,
     memory::{Memory, PageBuf, PageNumber},
+    message::{MessageId, OutgoingMessage, ReplyMessage},
 };
 
-pub trait Environment<E: Ext>: Default + Sized {
-    fn new() -> Self;
+pub const WAIT_TRAP_STR: &str = "wait";
+pub const EXIT_TRAP_STR: &str = "exit";
 
-    fn setup_and_run(
+pub enum TerminationReason {
+    Success,
+    Trap {
+        explanation: Option<&'static str>,
+        description: Option<String>,
+    },
+    Manual {
+        wait: bool,
+    },
```

Comment:
I don't think that's a good one, because I want to make accent on the reason of interrupting. There is actually 4 cases for future: `Success`, `Trap` (not handled by user), `Manual` and `Exit`, where `Exit` - some kind of program kill and `Manual` may be waiting case or custom panic without state rollback (extra syscall that I gonna provide after this PR merged). What do you think about it?

Reply:
Agree, let's leave it as is.